core/lib/multivm/src/versions/testonly/account_validation_rules.rs

```diff
@@ -0,0 +1,109 @@
+use assert_matches::assert_matches;
+use ethabi::Token;
+use zksync_types::{
+    fee::Fee, l2::L2Tx, transaction_request::TransactionRequest, Address, Eip712Domain, Execute,
+    L2ChainId, U256,
+};
+use zksync_vm_interface::{tracer::ViolatedValidationRule, InspectExecutionMode, VmInterfaceExt};
+
+use super::{
+    read_validation_test_contract, tester::VmTesterBuilder, ContractToDeploy, TestedVm,
+    TestedVmForValidation,
+};
+use crate::interface::TxExecutionMode;
+
+/// Checks that every limitation imposed on account validation results in an appropriate error.
+/// The actual misbehaviours are found in "validation-rule-breaker.sol".
+pub(crate) fn test_account_validation_rules<VM: TestedVm + TestedVmForValidation>() {
+    assert_matches!(test_rule::<VM>(0), None);
+    assert_matches!(
+        test_rule::<VM>(1),
+        Some(ViolatedValidationRule::TouchedDisallowedStorageSlots(_, _))
+    );
+    assert_matches!(
+        test_rule::<VM>(2),
+        Some(ViolatedValidationRule::CalledContractWithNoCode(_))
+    );
+
+    // Disallowing gasleft is difficult because it isn't immediately followed
+    // by far call in EraVM bytecode.
+    /*assert_matches!(
+        test_rule::<VM>(100),
+        Some(ViolatedValidationRule::TouchedDisallowedContext)
+    );*/
```

Comment:
I don't understand how this comment and the commented code relate to each other. AFAIU, `TouchedDisallowedContext` is raised by the new VM on `ContextMeta` opcode, which differs from `ErgsLeft`. Is it impossible to trigger `ContextMeta` opcode execution from Solidity? (Even if so, the comment is still misleading, and commented code should be either restored or removed.)

Reply:
The comment should be expanded a bit. The issue is that I want to test gasleft but we currently allow gasleft because there is no easy way to forbid it.
Using gasleft would also trigger TouchedDisallowedContext.